### DISCUSSION TITLE
Classes in single table inheritance with computed discriminator try to insert a value to it and fail

### DISCUSSION THREAD

**Discussion by ljmc-github** [sqlalchemy/sqlalchemy#11972] [Usage Questions] [answered]
I have a single table polymorphism mapped to some classes, the discriminator comes from a computed column (generated always stored).

When inserting, sqlalchemy tries to push the `polymorphic_identity` value to the generated discriminator and fails.

I've inspected the reflection event and saw the column is reflected as computed, but that doesn't seem to prevent the inserts like in `server_default` or `Computed` columns.

I've tried using the properties dict to make the discriminator a `column_property`, an explicit `Computed` and a few more things none worked so far.

I have read through https://github.com/sqlalchemy/sqlalchemy/discussions/11195, but I'm unsure if/how the same solution would apply as I'm using imperative mapping.

MCRE:

```python
from dataclasses import dataclass

from sqlalchemy import MetaData, create_engine, text
from sqlalchemy.orm import Session, registry

metadata = MetaData()
mapper_registry = registry()

engine = create_engine("sqlite://", echo=True)

with engine.connect() as con:
    con.execute(text(
            """\
CREATE TABLE fish (
    id INT PRIMARY KEY,
    weight INT NOT NULL,
    category TEXT GENERATED ALWAYS AS (CASE WHEN weight < 5 THEN 'small' ELSE 'big' END) STORED
);
"""
    ))
    con.execute(text("INSERT INTO fish (weight) VALUES (1), (6), (4), (7);"))
    con.commit()

metadata.reflect(engine)

@dataclass
class Fish:
    weight: int

@dataclass
class BigFish(Fish): ...

@dataclass
class SmallFish(Fish): ...

table_fish = metadata.tables["fish"]
mapper_fish = mapper_registry.map_imperatively(
    Fish,
    table_fish,
    polymorphic_on="category",
    polymorphic_abstract=True,
)
mapper_registry.map_imperatively(BigFish, inherits=mapper_fish, polymorphic_identity="big")
mapper_registry.map_imperatively(SmallFish, inherits=mapper_fish, polymorphic_identity="small")

with Session(engine) as session:
    session.add(SmallFish(weight=2))
    session.flush()

# OperationalError: (sqlite3.OperationalError) cannot INSERT into generated column "category"
# [SQL: INSERT INTO fish (weight, category) VALUES (?, ?)]
# [parameters: (2, 'small')]
# (Background on this error at: https://sqlalche.me/e/20/e3q8)
```

NB. this is a dummy example, real case involves a polymorphic table of entities whose role depends which of several foreign key columns to different tables is populated, but it illustrates well enough I think, happy to produce a more thorough example if necessary

**Comment by zzzeek** [comment, marked as answer]
for a read only [polymorphic_on](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.Mapper.params.polymorphic_on) you can set the parameter as a SQL expression that's not an immediate mapped column - the least intrusive SQL wise would be a type_coerce:

```py
from sqlalchemy import type_coerce, String
mapper_fish = mapper_registry.map_imperatively(
    Fish,
    table_fish,
    polymorphic_on=type_coerce(table_fish.c.category, String),
    polymorphic_abstract=True,
)
```

two problems with your example:

1. `id INT PRIMARY KEY` will not autoincrement in SQLite and the example fails - needs to be `id INTEGER PRIMARY KEY`
2. there's an architectural flaw in this design overall, in that I could write this:

```py
with Session(engine) as session:
    session.add(BigFish(weight=2))
    session.flush()
```

now you have an object that should be impossible.  you might want to use [@validates](https://docs.sqlalchemy.org/en/20/orm/mapped_attributes.html#sqlalchemy.orm.validates) to guard against this.

proof of concept

```py
from dataclasses import dataclass

from sqlalchemy import MetaData, create_engine, text, type_coerce, String
from sqlalchemy.orm import Session, registry

metadata = MetaData()
mapper_registry = registry()

engine = create_engine("sqlite://", echo='debug')

with engine.connect() as con:
    con.execute(text(
            """\
CREATE TABLE fish (
    id INTEGER PRIMARY KEY,
    weight INT NOT NULL,
    category TEXT GENERATED ALWAYS AS (CASE WHEN weight < 5 THEN 'small' ELSE 'big' END) STORED
);
"""
    ))
    con.execute(text("INSERT INTO fish (weight) VALUES (1), (6), (4), (7);"))
    con.commit()

metadata.reflect(engine)


@dataclass
class Fish:
    weight: int


@dataclass
class BigFish(Fish): ...

@dataclass
class SmallFish(Fish): ...

table_fish = metadata.tables["fish"]
mapper_fish = mapper_registry.map_imperatively(
    Fish,
    table_fish,
    polymorphic_on=type_coerce(table_fish.c.category, String),
    polymorphic_abstract=True,
)
mapper_registry.map_imperatively(BigFish, inherits=mapper_fish, polymorphic_identity="big")
mapper_registry.map_imperatively(SmallFish, inherits=mapper_fish, polymorphic_identity="small")

with Session(engine) as session:
    session.add(SmallFish(weight=2))
    session.flush()


```

**Reply by ljmc-github** [reply]
thank you very much, `polymorphic_on=type_coerce(...)` works perfectly

about your concerns:
1. sorry for the bug in the SQL, I originally wrote this for postgres and before doing a sqlite version, I must have copy pasted a bad intermediate version here
2. there are guard in the real life case exactly against that kind of invalid case
